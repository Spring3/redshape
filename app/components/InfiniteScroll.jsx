import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import throttle from 'lodash/throttle';

class InfiniteScroll extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isLoading: false
    };

    this.throttledScrollHandler = throttle(this.onScroll, 100);
  }

  componentWillUnmount() {
    if (this.props.container) {
      this.props.container.removeEventListener('scroll', this.throttledScrollHandler);
    }
  }

  componentDidUpdate(oldProps, oldState) {
    if (oldProps.isEnd !== this.props.isEnd && this.props.isEnd) {
      this.setState({
        isLoading: false
      });
    }

    if (oldProps.hasMore !== this.props.hasMore && !this.props.hasMore) {
      if (this.props.container) {
        this.props.container.removeEventListener('scroll', this.throttledScrollHandler);
      }
    }

    if (oldState.isLoading !== this.state.isLoading && !this.state.isLoading) {
      if (this.props.container && this.props.container.scrollTop <= this.props.container.scrollHeight) {
        this.throttledScrollHandler()
      }
    }

    if (oldProps.hasMore !== this.props.hasMore && this.props.hasMore) {
      if (this.state.isLoading) {
        this.setState({
          isLoading: false
        });
      }
      if (this.props.container) {
        this.props.container.addEventListener('scroll', this.throttledScrollHandler);
        if (this.props.immediate && this.props.container.scrollTop <= this.props.container.scrollHeight) {
          this.throttledScrollHandler();
        }
      }
    }
  }

  onScroll = (event = {}) => {
    const { load, threshold, hasMore, container } = this.props;
    const { isLoading } = this.state;
    const target = event.target || container;
    const scrolledToBottom = Math.ceil(target.scrollTop + target.clientHeight) >= target.scrollHeight;

    if (!isLoading && scrolledToBottom && hasMore) {
      if (target.scrollTop <= target.scrollHeight || (target.scrollTop % target.scrollHeight) >= threshold) {
        this.setState({
          isLoading: true
        }, () => load());
      }
    }
  };

  render() {
    const { children, loadIndicator } = this.props;
    const { isLoading } = this.state;
    return (
      <Fragment>
        {children}
        {isLoading && loadIndicator}
      </Fragment>
    );
  }
};

InfiniteScroll.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element.isRequired,
    PropTypes.arrayOf(PropTypes.element).isRequired
  ]).isRequired,
  hasMore: PropTypes.bool.isRequired,
  isEnd: PropTypes.bool.isRequired,
  load: PropTypes.func.isRequired,
  container: PropTypes.element,
  loadIndicator: PropTypes.node,
  threshold: PropTypes.number,
  immediate: PropTypes.bool
};

InfiniteScroll.defaultProps = {
  loadIndicator: null,
  container: null,
  threshold: 250,
  immediate: false
};

export default InfiniteScroll;
