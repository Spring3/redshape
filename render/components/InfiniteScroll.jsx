import React, { Fragment, Component } from 'react';
import PropTypes from 'prop-types';
import throttle from 'lodash/throttle';

class InfiniteScroll extends Component {
  constructor(props) {
    super(props);

    this.state = {
      isLoading: false
    };

    this.throttledScrollHandler = throttle(this.onScroll, 100);
  }

  componentWillMount() {
    if (this.props.container && this.props.immediate) {
      this.throttledScrollHandler();
    }
  }

  componentWillUnmount() {
    if (this.props.container) {
      this.props.container.removeEventListener('scroll', this.throttledScrollHandler);
    }
  }

  shouldLoad = () => {
    const { container } = this.props;
    if (container === window) {
      return (container.innerHeight + container.scrollY) <= document.body.scrollHeight;
    } else {
      return container.scrollTop <= container.scrollHeight;
    }
  }

  componentDidUpdate(oldProps, oldState) {
    if (oldProps.isEnd !== this.props.isEnd && this.props.isEnd) {
      this.setState({
        isLoading: false
      });
    }

    if (oldProps.hasMore !== this.props.hasMore && !this.props.hasMore) {
      if (this.props.container) {
        this.props.container.removeEventListener('scroll', this.throttledScrollHandler);
      }
    }

    if (oldState.isLoading !== this.state.isLoading && !this.state.isLoading) {
      if (this.props.container && this.shouldLoad()) {
        this.throttledScrollHandler()
      }
    }

    if (oldProps.hasMore !== this.props.hasMore && this.props.hasMore) {
      if (this.state.isLoading) {
        this.setState({
          isLoading: false
        });
      }
      if (this.props.container) {
        this.props.container.addEventListener('scroll', this.throttledScrollHandler);
        if (this.props.immediate && this.shouldLoad()) {
          this.throttledScrollHandler();
        }
      }
    }
  }

  onScroll = (event = {}) => {
    const { load, threshold, hasMore, container } = this.props;
    const { isLoading } = this.state;
    const target = event.target || container;
    const scrolledToBottom = container === window
     ? (window.innerHeight + window.scrollY) >= document.body.scrollHeight
     : Math.ceil(target.scrollTop + target.clientHeight) >= target.scrollHeight;
    // if does not have an overflow, or scrolled to bottom
    if (!isLoading && scrolledToBottom && hasMore) {
      if (this.shouldLoad() || (target.scrollTop % target.scrollHeight) >= threshold) {
        this.setState({
          isLoading: true
        }, () => load());
      }
    }
  };

  render() {
    const { children, loadIndicator } = this.props;
    const { isLoading } = this.state;
    return (
      <Fragment>
        {children}
        {isLoading && loadIndicator}
      </Fragment>
    );
  }
};

InfiniteScroll.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.element.isRequired,
    PropTypes.arrayOf(PropTypes.element).isRequired
  ]).isRequired,
  hasMore: PropTypes.bool.isRequired,
  isEnd: PropTypes.bool.isRequired,
  load: PropTypes.func.isRequired,
  container: PropTypes.object,
  loadIndicator: PropTypes.node,
  threshold: PropTypes.number,
  immediate: PropTypes.bool
};

InfiniteScroll.defaultProps = {
  loadIndicator: null,
  container: null,
  threshold: 250,
  immediate: false
};

export default InfiniteScroll;
