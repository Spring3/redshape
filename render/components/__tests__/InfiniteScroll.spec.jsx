import React from 'react';
import { render, fireEvent, cleanup } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';

import InfiniteScroll from '../InfiniteScroll';

describe('InfiniteScroll component', () => {
  afterEach(cleanup);
  it('should set the scroll event handler once initialized', () => {
    const containerMock = {
      addEventListener: jest.fn(),
      removeEventListener: jest.fn()
    };

    const { queryByText, rerender } = render(
      <InfiniteScroll
        immediate={true}
        isEnd={false}
        hasMore={false}
        load={() => {}}
        loadIndicator={<span>Loading...</span>}
        container={containerMock}
      >
        <div>Hello</div>
        <div>World</div>
      </InfiniteScroll>
    );

    rerender(
      <InfiniteScroll
        immediate={true}
        isEnd={false}
        hasMore={true}
        load={() => {}}
        loadIndicator={<span>Loading...</span>}
        container={containerMock}
      >
        <div>Hello</div>
        <div>World</div>
      </InfiniteScroll>
    );

    expect(containerMock.addEventListener).toHaveBeenCalled();
    expect(queryByText('Hello')).toBeDefined();
    expect(queryByText('World')).toBeDefined();
  });

  it('should remove the scroll event handle before unmount', () => {
    const containerMock = {
      addEventListener: jest.fn(),
      removeEventListener: jest.fn()
    };
    const { rerender, unmount } = render(
      <InfiniteScroll
        immediate={true}
        isEnd={false}
        hasMore={false}
        load={() => {}}
        loadIndicator={<span>Loading...</span>}
        container={containerMock}
      >
        <div>Hello</div>
        <div>World</div>
      </InfiniteScroll>
    );
    rerender(
      <InfiniteScroll
        immediate={true}
        isEnd={false}
        hasMore={true}
        load={() => {}}
        loadIndicator={<span>Loading...</span>}
        container={containerMock}
      >
        <div>Hello</div>
        <div>World</div>
      </InfiniteScroll>
    );
    expect(containerMock.addEventListener).toHaveBeenCalled();
    unmount();
    expect(containerMock.removeEventListener).toHaveBeenCalled();
  });

  it('should remove the scroll handler if hasMore is set to false', () => {
    const containerMock = {
      addEventListener: jest.fn(),
      removeEventListener: jest.fn()
    };
    const { rerender } = render(
      <InfiniteScroll
        immediate={true}
        isEnd={false}
        hasMore={true}
        load={() => {}}
        loadIndicator={<span>Loading...</span>}
        container={containerMock}
      >
        <div>Hello</div>
        <div>World</div>
      </InfiniteScroll>
    );
    rerender(
      <InfiniteScroll
        immediate={true}
        isEnd={false}
        hasMore={false}
        load={() => {}}
        loadIndicator={<span>Loading...</span>}
        container={containerMock}
      >
        <div>Hello</div>
        <div>World</div>
      </InfiniteScroll>
    );
    expect(containerMock.removeEventListener).toHaveBeenCalled();
  });

  describe('onScroll function', () => {
    it('should set isLoading to true and call load function if scrolled to bottom', (done) => {
      const containerMock = {
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        scrollTop: 0,
        scrollHeight: 300,
        clientHeight: 300
      };
      const loadFn = jest.fn();
      const { queryByText, baseElement } = render(
        <InfiniteScroll
          immediate={true}
          isEnd={false}
          hasMore={true}
          load={loadFn}
          loadIndicator={<span>Loading...</span>}
          container={containerMock}
        >
          <div>Hello</div>
          <div>World</div>
        </InfiniteScroll>
      );
      fireEvent.scroll(baseElement, {});
      setTimeout(() => {
        expect(queryByText('Loading...')).toBeDefined();
        expect(loadFn).toHaveBeenCalled();
        done();
      }, 1);
    });
  });
});
