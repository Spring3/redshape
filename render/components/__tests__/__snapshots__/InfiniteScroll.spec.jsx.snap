// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`InfiniteScroll component should match the snapshot 1`] = `
Array [
  <div>
    Hello
  </div>,
  <div>
    World
  </div>,
]
`;
